flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java

```diff
@@ -0,0 +1,299 @@
+import io.fabric8.kubernetes.client.informers.SharedIndexInformer;
+import io.fabric8.kubernetes.client.informers.SharedInformerFactory;
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+
+/** Base class for shared watcher based on {@link SharedIndexInformer}. */
+public abstract class KubernetesSharedInformer<
+                T extends HasMetadata, TList extends KubernetesResourceList<T>, R>
+        implements KubernetesSharedWatcher<R> {
+
+    protected final Logger log = LoggerFactory.getLogger(getClass());
+
+    private final NamespacedKubernetesClient client;
+    private final SharedIndexInformer<T> sharedIndexInformer;
+    private final Function<T, R> eventWrapper;
+
+    private final ExecutorService watchingExecutor;
+
+    private final Object processorsLock = new Object();
+
+    @GuardedBy("processorsLock")
+    private final Map<String, Processor> watchingProcessors = new HashMap<>();
+
+    public KubernetesSharedInformer(
+            NamespacedKubernetesClient client,
+            Class<T> apiTypeClass,
+            Class<TList> apiListTypeClass,
+            Map<String, String> labels,
+            Function<T, R> eventWrapper) {
+        this.client = client;
+        SharedInformerFactory informerFactory =
+                client.informers(
+                        Executors.newSingleThreadExecutor(
+                                new ExecutorThreadFactory("KubernetesClient-Informer")));
+        if (!CollectionUtil.isNullOrEmpty(labels)) {
+            informerFactory.withLabels(labels);
+        }
+        this.sharedIndexInformer =
+                informerFactory.sharedIndexInformerFor(apiTypeClass, apiListTypeClass, 30000);
+        this.sharedIndexInformer.addEventHandler(
+                new ResourceEventHandler<T>() {
+                    @Override
+                    public void onAdd(T obj) {
+                        notify(obj.getMetadata().getName());
+                    }
+
+                    @Override
+                    public void onUpdate(T oldObj, T newObj) {
+                        notify(newObj.getMetadata().getName());
+                    }
+
+                    @Override
+                    public void onDelete(T obj, boolean deletedFinalStateUnknown) {
+                        notify(obj.getMetadata().getName());
+                    }
+
+                    private void notify(String name) {
```

Comment:
It's a normal phenomenon if watching after the resource created. It doesn't matter since the handler will always receive an ADDED event.

Processor will trigger an event by itself before starting to run, with which it will retrieve the current Resource.

Reply:
Do we have to strictly restrict that one single get operation must be guarded in lock?

Reply:
The IDE complains that.